File: BackJoon/gold3/P6087.py
# 레이저 통신

import sys

input = sys.stdin.readline

def solution():
    col, row = map(int, input().split())
    maps = []
    target = []
    for i in range(row):
        maps.append(list(input().replace("\n", '')))
        for j in range(col):
            if maps[i][j] == 'C':
                target.append((i, j))

    start, end = target

    q = [(*start, 0), (*start, 1), (*start, 2), (*start, 3)]

    dp = [[[0, 0, 0, 0] for _ in range(col)] for _ in range(row)]

    dp[start[0]][start[1]][0] = 1
    dp[start[0]][start[1]][1] = 1
    dp[start[0]][start[1]][2] = 1
    dp[start[0]][start[1]][3] = 1


    while q:
        i, j, d_t = q.pop(0)

        for d_i, m in enumerate([(1, 0), (0, 1), (-1, 0), (0, -1)]): #down, right, up, left
            m_i = i + m[0]
            m_j = j + m[1]

            if 0 <= m_i and m_i < row and 0 <= m_j and m_j < col:
                if maps[m_i][m_j] != '*':
                    if d_t % 2 == d_i % 2: #같은 방향
                        if dp[m_i][m_j][d_i] == 0:
                            dp[m_i][m_j][d_i] = dp[i][j][d_t]
                            q.append((m_i, m_j, d_i))
                        else:
                            if dp[m_i][m_j][d_i] > dp[i][j][d_t]:
                                dp[m_i][m_j][d_i] = dp[i][j][d_t]
                                q.append((m_i, m_j, d_i))
                    elif d_t % 2 != d_i % 2: #다른 방향
                        if dp[m_i][m_j][d_i] == 0:
                            dp[m_i][m_j][d_i] = dp[i][j][d_t] + 1
                            q.append((m_i, m_j, d_i))
                        else:
                            if dp[m_i][m_j][d_i] > dp[i][j][d_t] + 1:
                                dp[m_i][m_j][d_i] = dp[i][j][d_t] + 1
                                q.append((m_i, m_j, d_i))

    min_value =214700000
    for i in dp[end[0]][end[1]]:
        if i != 0 and min_value > i:
            min_value = i

    print(min_value - 1)

solution()

File: BackJoon/gold3/P6087.cpp
#include <iostream>
#include <string>
#include <vector>
#include <queue>

using namespace std;

int ROW, COL;

struct point {
	int row, col;
	int direction;

	point(int row, int col) : row(row), col(col) {
		direction = -1;
	}

	point(const point& p, int direction) : row(p.row), col(p.col), direction(direction) {}

	point(int row, int col, int direction) : row(row), col(col), direction(direction) {}
};

template <typename T>
void print_map(T** map, int direction = 0) {
	cout << "-----------------------------------\n";
	for (int i = 0; i < ROW; i++) {
		for (int j = 0; j < COL; j++) {
			string tmp = to_string(map[i][j][direction]);
			if (tmp.length() == 3) {
				cout << tmp;
			}
			else if (tmp.length() == 2) {
				cout << " " << tmp;
			}
			else if (tmp.length() == 1) {
				cout << "  " << tmp;
			}
			else {
				cout << "  X";
			}
		}
		cout << "\n";
	}
}

int solution(char** map, vector<point> ps) {

	// down, right, up, left
	point move[4] = { point(1, 0), point(0, 1),point(-1, 0), point(0, -1) };

	int*** mirror_map = new int** [ROW];
	for (int i = 0; i < ROW; i++) {
		mirror_map[i] = new int*[COL];
		for (int j = 0; j < COL; j++) {
			mirror_map[i][j] = new int[4];
			for (int k = 0; k < 4; k++) {
				mirror_map[i][j][k] = INT32_MAX;
			}
		}
	}

	queue<point> bfs_q;
	//for (int i = 0; i < 4; i++) {
	//	bfs_q.push(point(ps[0], i));
	//	mirror_map[ps[0].row][ps[0].col][i] = 0;
	//}
	bfs_q.push(point(ps[0]));
	for (int i = 0; i < 4; i++) {
		mirror_map[ps[0].row][ps[0].col][i] = 0;
	}
	while (!bfs_q.empty()) {
		point cur = bfs_q.front();
		bfs_q.pop();

		for (int i = 0; i < 4; i++) {
			int mov_r = cur.row + move[i].row;
			int mov_c = cur.col + move[i].col;

			if (0 <= mov_r && mov_r < ROW && 0 <= mov_c && mov_c < COL) {
				if (map[mov_r][mov_c] != '*') {
					//bfs_q.push(point(mov_r, mov_c, i));

					if (cur.direction == -1) { //시작 지점
						bfs_q.push(point(mov_r, mov_c, i));
						mirror_map[mov_r][mov_c][i] = mirror_map[cur.row][cur.col][0];
					}
					else if (cur.direction % 2 != i % 2) { //직각 방향으로 움직일경우
						if (mirror_map[mov_r][mov_c][i] > mirror_map[cur.row][cur.col][cur.direction] + 1) {
							bfs_q.push(point(mov_r, mov_c, i));
							mirror_map[mov_r][mov_c][i] = mirror_map[cur.row][cur.col][cur.direction] + 1;
						}
					}
					else { // 평행 방향으로 움직일 경우
						if (mirror_map[mov_r][mov_c][i] > mirror_map[cur.row][cur.col][cur.direction]) {
							bfs_q.push(point(mov_r, mov_c, i));
							mirror_map[mov_r][mov_c][i] = mirror_map[cur.row][cur.col][cur.direction];
						}
					}
				}
			}
		}
	}

	int min_v = INT32_MAX;
	for (int i = 0; i < 4; i++) {
		//cout << mirror_map[ps[1].row][ps[1].col][i] << " ";
		if (mirror_map[ps[1].row][ps[1].col][i] != -1) {
			min_v = min(min_v, mirror_map[ps[1].row][ps[1].col][i]);
		}
	}
	//cout << "\n";

	/*for (int i = 0; i < 4; i++) {
		print_map(mirror_map, i);
	}*/

	return min_v;
}

int main() {
	cin.tie(NULL)->sync_with_stdio(false);

	cin >> COL >> ROW;

	char** map = new char*[ROW];

	vector<point> c_points;

	for (int i = 0; i < ROW; i++) {
		map[i] = new char[COL];
		string tmp;
		cin >> tmp;
		for (int j = 0; j < COL; j++) {
			map[i][j] = tmp[j];
			if (map[i][j] == 'C') {
				c_points.push_back(point(i, j));
			}
		}
	}

	cout << solution(map, c_points) << "\n";

	return 0;
}